native/cpp/LSPWorkspace.cc

```diff
@@ -0,0 +1,224 @@
+    }
+    Pol::Plib::replace_packages();
+    Pol::Plib::check_package_deps();
+    return env.Undefined();
+  }
+  catch ( const std::exception& ex )
+  {
+    Napi::Error::New( env, ex.what() ).ThrowAsJavaScriptException();
+  }
+  catch ( ... )
+  {
+    Napi::Error::New( env, "Unknown Error" ).ThrowAsJavaScriptException();
+  }
+
+  return Napi::Value();
+}
+
+Napi::Value LSPWorkspace::Open( const Napi::CallbackInfo& info )
+{
+  auto env = info.Env();
+
+  if ( info.Length() < 1 || !info[0].IsString() )
+  {
+    Napi::TypeError::New( env, Napi::String::New( env, "Invalid arguments" ) )
+        .ThrowAsJavaScriptException();
+  }
+
+  auto pathname = info[0].As<Napi::String>().Utf8Value();
+  auto extension = std::filesystem::path( pathname ).extension().string();
+  Pol::Clib::mklowerASCII( extension );
+  bool is_module = extension.compare( ".em" ) == 0;
+  _cache.emplace( std::make_pair( pathname, LSPDocument( *this, pathname, is_module ) ) );
+  return info.Env().Undefined();
+}
+
+Napi::Value LSPWorkspace::Close( const Napi::CallbackInfo& info )
+{
+  auto env = info.Env();
+
+  if ( info.Length() < 1 || !info[0].IsString() )
+  {
+    Napi::TypeError::New( env, Napi::String::New( env, "Invalid arguments" ) )
+        .ThrowAsJavaScriptException();
+  }
+
+  auto pathname = info[0].As<Napi::String>();
+  auto itr = _cache.find( pathname );
+  if ( itr != _cache.end() )
+  {
+    _cache.erase( itr );
+    return Napi::Boolean::New( env, true );
+  }
+  return Napi::Boolean::New( env, false );
+}
+
+Napi::Value LSPWorkspace::Precompile( const Napi::CallbackInfo& info )
+{
+  auto env = info.Env();
+
+  if ( info.Length() < 1 || !info[0].IsString() )
+  {
+    Napi::TypeError::New( env, Napi::String::New( env, "Invalid arguments" ) )
+        .ThrowAsJavaScriptException();
+  }
+
+  auto pathname = info[0].As<Napi::String>();
+  auto itr = _cache.find( pathname );
+  if ( itr == _cache.end() )
+  {
+    Napi::Error::New( env, Napi::String::New( env, "Document not opened" ) )
+        .ThrowAsJavaScriptException();
+  }
+  auto& document = itr->second;
+  try
+  {
+    document.precompile();
+    return env.Undefined();
+  }
+  catch ( const std::exception& ex )
+  {
+    Napi::Error::New( env, ex.what() ).ThrowAsJavaScriptException();
+  }
+  catch ( ... )
+  {
+    Napi::Error::New( env, "Unknown Error" ).ThrowAsJavaScriptException();
+  }
+  return Napi::Value();
+}
+
+Napi::Value LSPWorkspace::Diagnostics( const Napi::CallbackInfo& info )
+{
+  auto env = info.Env();
+
+  if ( info.Length() < 1 || !info[0].IsString() )
+  {
+    Napi::TypeError::New( env, Napi::String::New( env, "Invalid arguments" ) )
+        .ThrowAsJavaScriptException();
+  }
+
+  auto pathname = info[0].As<Napi::String>();
+  auto itr = _cache.find( pathname );
+  if ( itr == _cache.end() )
+  {
+    Napi::Error::New( env, Napi::String::New( env, "Document not opened" ) )
+        .ThrowAsJavaScriptException();
+  }
+  auto& document = itr->second;
```

Comment:
You might consider extracting this as a method. It's line for line what is done in Precompile(...)

Reply:
yuup absolutely! I noticed that I am now adding more methods (tokens, dependees) and they will all have the same "preamble". I will refactor once I have more methods to properly get the right api